File: core/query/graph_scale_traversal_engine.py
from __future__ import annotations

from collections import deque

from typing import Any, Dict, List


MAX_GRAPH_VISITS = 10000


def traverse_large_graph(
    graph: Dict[str, Any],
    start: str,
) -> Dict[str, Any]:

    adjacency = {}

    for edge in graph.get("edges", []):

        adjacency.setdefault(
            edge["from"],
            [],
        ).append(edge["to"])

    queue = deque([start])

    visited = []

    seen = set()

    while queue and len(visited) < MAX_GRAPH_VISITS:

        node = queue.popleft()

        if node in seen:
            continue

        seen.add(node)

        visited.append(node)

        for nxt in sorted(adjacency.get(node, [])):
            queue.append(nxt)

    return {
        "visited": visited,
        "count": len(visited),
        "bounded": True,
    }

File: src/query/graphScaleTraversalEngine.ts
/**
 * Converted from Python: core/query/graph_scale_traversal_engine.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";

export let MAX_GRAPH_VISITS: any = 10000;
export function traverseLargeGraph(graph: any, start: any): any {
  var adjacency: Record<string, any> = {};
  var edge: any;
  for (edge of py.iter(py.get(graph, "edges", []))) {
    py.listAppend(py.setdefault(adjacency, py.at(edge, "from"), []), py.at(edge, "to"));
  }
  var queue: any = py.deque([start]);
  var visited: any[] = [];
  var seen: Set<any> = new Set();
  while ((py.truthy(queue) && (py.len(visited) < MAX_GRAPH_VISITS))) {
    var node: any = py.popleft(queue);
    if (py.contains(seen, node)) {
      continue;
    }
    py.setAdd(seen, node);
    py.listAppend(visited, node);
    var nxt: any;
    for (nxt of py.iter(py.sorted(py.get(adjacency, node, [])))) {
      py.listAppend(queue, nxt);
    }
  }
  return {"visited": visited, "count": py.len(visited), "bounded": true};
}
